sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -158,48 +175,63 @@ private static void WriteMemory(MemoryStream stream, ReadOnlyMemory<byte> memory
         /// ]]></code>
         /// </example>
         /// <remarks> See error response code and error response message for more detail. </remarks>
-        public virtual Response<UploadLogsResult> Upload<T>(string ruleId, string streamName, IEnumerable<T> logEntries, CancellationToken cancellationToken = default)
+        public virtual Response<UploadLogsResult> Upload<T>(string ruleId, string streamName, IEnumerable<T> logEntries, UploadLogsOptions options = null, CancellationToken cancellationToken = default)
         {
             Argument.AssertNotNullOrEmpty(ruleId, nameof(ruleId));
             Argument.AssertNotNullOrEmpty(streamName, nameof(streamName));
             Argument.AssertNotNullOrEmpty(logEntries, nameof(logEntries));
 
             using var scope = ClientDiagnostics.CreateScope("LogsIngestionClient.Upload");
-            scope.Start();
 
             RequestContext requestContext = GenerateRequestContext(cancellationToken);
             Response response = null;
             List<UploadLogsError> errors = new List<UploadLogsError>();
+            //TODO: use options.Serializer
             try
             {
+                scope.Start();
+                // A list of tasks that are currently executing which will
+                // always be smaller than MaxWorkerCount
+                List<Task<RunningTask<T>>> runningTasks = new();
+                // Partition the stream into individual blocks
                 foreach (BatchedLogs<T> batch in Batch(logEntries))
                 {
-                    //TODO: catch errors and correlate with Batch.start
-                    using HttpMessage message = CreateUploadRequest(ruleId, streamName, batch.LogsData, "gzip", requestContext);
-                    response = _pipeline.ProcessMessage(message, requestContext, cancellationToken);
-                    if (response.Status != 204) // if any error is thrown log it
+                    // Because we are uploading in sequence, wait for each batch to upload before starting the next batch
+                    RunningTask<T> task = CommitBatchListSyncOrAsync(
+                        batch,
+                        ruleId,
+                        streamName,
+                        false,
+                        cancellationToken).EnsureCompleted();
+                }
+
+                // Process all errors after tasks are done to determine status
+                // Will run on a single thread
+                foreach (Task<RunningTask<T>> task in runningTasks)
```

Comment:
We did not add anything to the runningTasks list? Did you intend to insert to it in the foreach loop?

Reply:
It looks like it didn't get copied over from the async implementation - https://github.com/Azure/azure-sdk-for-net/pull/31074/files#diff-6767d7871ae91cf5eb10fa3e4f24a9d5bbd9f9dbcc22bcbd72be0eabe315c103R290

It would be great to factor this out into a separate private method that can be called from both the sync and async public methods.

Reply:
oops! forgot to add to the list.

Reply:
Chatted with @nisha-bhatia - parallel uploads are not supported when using the sync method, hence the different implementations.

Reply:
Thinking about this more, I think we should still try to factor out the implementation into a separate private method. We would just need to branch in a few places on `isAsync` to avoid awaiting tasks in the sync path. In the sync path, the tasks would all be completed whenever we get them back.

Reply:
I have a helper method now that handles both sync/async